### PR TITLE
HW-53466 53467 prepaid cards

### PR DESCRIPTION
**Pull request by azakrevska-epam**

### CODE REVIEW

File: core/src/test/java/com/hyperwallet/android/transfermethod/HyperwalletListPrepaidCardReceiptsTest.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+import org.mockito.junit.MockitoRule;
+import org.robolectric.RobolectricTestRunner;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import okhttp3.mockwebserver.RecordedRequest;
+
+@RunWith(RobolectricTestRunner.class)
+public class HyperwalletListPrepaidCardReceiptsTest {
```
Comment (peter-joseph) [inline core/src/test/java/com/hyperwallet/android/transfermethod/HyperwalletListPrepaidCardReceiptsTest.java]: I noticed that all test has query param, please choose a test case that demonstrate null query parameter 
and moreover choose a test case that uses `createdBefore` and `createdAfter` query parameters, this is important since prepaid paging is only using this attributes

**tip** on testing on dates create a string date parse with our date util, and generate dateobjects from that so that, when you assert the URL path query params `createdBefore` and `createdAfter` will always have same result.

File: core/src/main/java/com/hyperwallet/android/Hyperwallet.java [hunk 2]
```diff
@@ -722,6 +722,44 @@ public void listUserReceipts(@NonNull final ReceiptQueryParam receiptQueryParam,
...
+     *
+     * <p>This function will request a new authentication token via {@link HyperwalletAuthenticationTokenProvider}
+     * if the current one is expired or about to expire.</p>
+     *
+     * @param prepaidCardToken  the token for prepaid card
+     * @param receiptQueryParam the filtering criteria
+     * @param listener          the callback handler of responses from the Hyperwallet platform; must not be null
+     */
+    public void listPrepaidCardReceipts(@NonNull final String prepaidCardToken,
+            @NonNull final ReceiptQueryParam receiptQueryParam,
```
[the remark this answers is not in the document]
Comment (azakrevska-epam) [reply core/src/main/java/com/hyperwallet/android/Hyperwallet.java]: @peter-joseph Thank you, done

File: core/src/test/java/com/hyperwallet/android/transfermethod/HyperwalletListPrepaidCardReceiptsTest.java (hunk 1 above)
Comment (azakrevska-epam) [reply core/src/test/java/com/hyperwallet/android/transfermethod/HyperwalletListPrepaidCardReceiptsTest.java]: @peter-joseph Could you take a look on the added tests?

File: core/src/main/java/com/hyperwallet/android/Hyperwallet.java [hunk 3]
```diff
@@ -722,6 +722,44 @@ public void listUserReceipts(@NonNull final ReceiptQueryParam receiptQueryParam,
...
+     *
+     * <p>This function will request a new authentication token via {@link HyperwalletAuthenticationTokenProvider}
+     * if the current one is expired or about to expire.</p>
+     *
+     * @param prepaidCardToken  the token for prepaid card
+     * @param receiptQueryParam the filtering criteria
+     * @param listener          the callback handler of responses from the Hyperwallet platform; must not be null
+     */
+    public void listPrepaidCardReceipts(@NonNull final String prepaidCardToken,
+            @Nullable final ReceiptQueryParam receiptQueryParam,
```
Comment (fmattos-hw) [inline core/src/main/java/com/hyperwallet/android/Hyperwallet.java]: I think the query param should be actually @NonNull - I think if we don't pass it the query will return 204 - no content. If that's the case I think we should enforce it being passed in

File: core/src/main/java/com/hyperwallet/android/Hyperwallet.java [hunk 4]
```diff
@@ -722,6 +722,44 @@ public void listUserReceipts(@NonNull final ReceiptQueryParam receiptQueryParam,
...
+     * {@link HyperwalletAuthenticationTokenProvider#retrieveAuthenticationToken(HyperwalletAuthenticationTokenListener)},
+     * or an empty {@code List} if non exist.
+     *
+     * <p>The ordering and filtering of {@code HyperwalletReceipts} will be based on the criteria specified within
+     * the {@link ReceiptQueryParam} object, if it is not null. Otherwise the default ordering and
+     * filtering will be applied:</p>
+     *
+     * <ul>
+     * <li>Created Before: N/A</li>
+     * <li>Created After: N/A</li>
```
Comment (fmattos-hw) [inline core/src/main/java/com/hyperwallet/android/Hyperwallet.java]: @azakrevska-epam confirm with @peter-joseph - I think we need to pass some specific date, something like one year back or the api won't return more than 20 records. Please confirm this with Peter and make sure the docs reflect that for users and prepaid cards receipts

File: core/src/test/java/com/hyperwallet/android/transfermethod/HyperwalletListPrepaidCardReceiptsTest.java (hunk 1 above)
Comment (fmattos-hw) [reply core/src/test/java/com/hyperwallet/android/transfermethod/HyperwalletListPrepaidCardReceiptsTest.java]: I think we need to get on the same page on this param. I think that should be documented via javadocs and tests removed. I'd like to keep the tests classes as simple and small as possible, I don't want to add edge case scenarios to our test classes

File: core/src/main/java/com/hyperwallet/android/Hyperwallet.java (hunk 3 above)
Comment (azakrevska-epam) [reply core/src/main/java/com/hyperwallet/android/Hyperwallet.java]: @fmattos-hw Done

Comment (peter-joseph) [reply core/src/main/java/com/hyperwallet/android/Hyperwallet.java]: 204 is normal, that means there's no transaction in our platform default date limit
which is a year from today

the default before behaviour for PPC dates will be request date as start date and end date will be a year from request date (this is based on platform code)

Comment (peter-joseph) [reply core/src/main/java/com/hyperwallet/android/Hyperwallet.java]: I think this still need to be `@Nullable` so that integrators can still use the default behaviour of our server if they wish too..

File: core/src/main/java/com/hyperwallet/android/Hyperwallet.java (hunk 4 above)
Comment (peter-joseph) [reply core/src/main/java/com/hyperwallet/android/Hyperwallet.java]: **Prepaid card** context we have date limit protection so user can have dates **createdBefore** and **createdAfter** as they wish, except for **createdAfter** is in the future
check this documentation too in **Query** section
https://portal.hyperwallet.com/docs/api/v3/resources/receipts/prepaid-cards-list
PPC receipts has max boundary of response to a year

**User receipt** context
same with PPC **createdBefore** and **createdAfter** but the range is much smaller like in a week as a default if this are not specified, or else **limit** and **offset** will take over the paging

File: core/src/main/java/com/hyperwallet/android/Hyperwallet.java (hunk 3 above)
Comment (fmattos-hw) [reply core/src/main/java/com/hyperwallet/android/Hyperwallet.java]: yeah you're right @peter-joseph 
@azakrevska-epam , please make the query param @Nullable - let's keep it consistent with the other list methods

Comment (azakrevska-epam) [reply core/src/main/java/com/hyperwallet/android/Hyperwallet.java]: @fmattos-hw Done

File: core/src/main/java/com/hyperwallet/android/Hyperwallet.java (hunk 4 above)
Comment (azakrevska-epam) [reply core/src/main/java/com/hyperwallet/android/Hyperwallet.java]: @peter-joseph Where can I find these default values for createdBefore and createdAfter dates in the documentation for PPC and UserReceipt?

Comment (peter-joseph) [reply core/src/main/java/com/hyperwallet/android/Hyperwallet.java]: @azakrevska-epam it means `createdBefore` and `createdAfter` is optional, you don't need to specify the default value for each, just describe the format for it, those parameters seems self explanatory too.. and if we want to explain more its better to have the portal documentation explain it

**Comment by skoong** [comment]
Right base branch? `task/HW-53465-53468-rename-user-receipts` is already merged into `development` https://github.com/hyperwallet/hyperwallet-android-sdk/pull/44

**Comment by azakrevska-epam** [comment]
> Right base branch? `task/HW-53465-53468-rename-user-receipts` is already merged into `development` #44

@skoong I'm sorrry, it is typo - this is right branch with right commit

**Comment by peter-joseph** [comment]
@azakrevska-epam this branch is merging into a merged branch, should this be going directly to `development` branch?
